node/core/candidate-validation/src/lib.rs

```diff
@@ -666,11 +666,14 @@ impl ValidationBackend for ValidationHost {
 
 	async fn precheck_pvf(&mut self, pvf: Pvf) -> Result<Duration, PrepareError> {
 		let (tx, rx) = oneshot::channel();
+		// TODO: Why are we always returning a non-deterministic error? This causes the pre-check
+		// outcome to always be `Failed`, never `Invalid`, thus never triggering disputes according
+		// to comments in pvf/src/error.rs. Can we change this to return the actual error?
 		if let Err(_) = self.precheck_pvf(pvf, tx).await {
-			return Err(PrepareError::DidNotMakeIt)
+			return Err(PrepareError::IoErr)
```

Comment:
Looks like for pre-checking we don't trigger disputes, wanted to double-check that this is intended.

Reply:
Disputes are arisen over an available parablock, that is, dispute is a result of a negative outcome of running a PVF over a given parablock. Invalid PVF results in a rejection to upgrade that PVF (or a rejection to onboard a parachain, whichever is requested), but cannot result in a dispute as the parablock itself is not invalid. (Disclaimer: it's just my understanding of what's happening, correct me if I'm wrong). So yes, returning `Failed` but not `Invalid` here seems legit to me.

Reply:
It has nothing to do with disputes. The sequence is as follows:
1. A candidate containing an upgrade is accepted by the Runtime.
2. Validator polls each new activation for pending pvfs.
3. Fetches it, initiates prechecking.
4. `Valid` -> submit "valid" vote via unsigned transaction in the next block, `Invalid` -> submit "invalid" one, `Failed` -> don't submit anything.

Thus your TODO is correct. 

Getting a little deeper: it doesn't harm the security because pvf still requires a threshold of valid votes to be accepted. But it also means that invalid pvf will be always rejected by the timeout.

Reply:
Yes

Reply:
Sorry, I totally misred this code. We are already returning the `precheck_result` a couple lines down. I kind of tunnel-visioned and didn't see that. The error that was getting swallowed here was an IO communication error with the host.

Reply:
Yes, disregard, ioerr is about channel being dropped